src/Avalonia.Controls/ThemeVariantWindowIcon.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.IO;
+using Avalonia.Platform;
+
+namespace Avalonia.Controls;
+
+/// <summary>
+/// Represents two icons, one for use with a light theme and the other for use with a dark theme.
+/// </summary>
+/// <remarks>
+/// A window's icon can appear in multiple locations (e.g. in the window title bar and the platform taskbar), and these locations 
+/// can use different themes. This means that both light and dark icons can be visible at the same time.
+/// </remarks>
+public class ThemeVariantWindowIcon : WindowIcon
+{
+    private static IWindowIconImpl Extract(WindowIcon icon, string paramName) => icon switch
+    {
+        ThemeVariantWindowIcon => throw new ArgumentException($"{nameof(ThemeVariantWindowIcon)} may not recursively contain further {nameof(ThemeVariantWindowIcon)}.", paramName),
+        null => throw new ArgumentNullException(paramName),
+        _ => icon.PlatformImpl,
+    };
+
+    /// <inheritdoc cref="ThemeVariantWindowIcon"/>
+    public ThemeVariantWindowIcon(WindowIcon light, WindowIcon dark) : base(new Impl(Extract(light, nameof(light)), Extract(dark, nameof(dark))))
```

Comment:
We cannot have an API that limits only to two possible theme variants.

If you refer to the OS theme variant, rather than app theme, it also won't be the best API, as it adds confusion.

Reply:
This is indeed the OS theme variant, either light or dark. This concept is already present in Avalonia via `Avalonia.Platform.PlatformThemeVariant`.

[The name of this type was discussed earlier](https://github.com/AvaloniaUI/Avalonia/pull/12107#pullrequestreview-1520718667) and I'm open to suggestions for a different name.